File: 1129.shortest-path-with-alternating-colors.py
#
# @lc app=leetcode id=1129 lang=python3
#
# [1129] Shortest Path with Alternating Colors
#

# @lc code=start
from dataclasses import dataclass
from pathlib import Path
from typing import List, Tuple


class Solution:
    @dataclass
    class Path:
        to:int
        color:bool
    
    def make_graph(self,redEdges:List[List[int]],blueEdges:List[List[int]],n:int): 
        self.graph:List[List[self.Path]]=[ [] for _ in range(n) ]
        for edges in redEdges:
            self.graph[edges[0]].append(self.Path(to=edges[1],color=False))
            
        for edges in blueEdges:
            self.graph[edges[0]].append(self.Path(to=edges[1],color=True))

        
            
    def dfs(self, current, current_stack: List[int],distance: int = 0):
        if self.distance[current.to]> distance or self.distance[current.to]==-1:
            self.distance[current.to] = distance
        for neighbor in self.graph[current.to]:
            print(neighbor.to," . ",current_stack)
            
            if current.color is None or current.color != neighbor.color:
                if neighbor.to in current_stack:
                    return
                current_stack.append(current.to)
                self.dfs(neighbor, distance=distance + 1,current_stack=current_stack)
        
 
    
    def shortestAlternatingPaths(self, n: int, redEdges: List[List[int]], blueEdges: List[List[int]]) -> List[int]:
        self.distance = [-1 for _ in range(n)]
        self.make_graph(redEdges=redEdges,blueEdges=blueEdges,n=n)   
        self.dfs(self.Path(0,None),current_stack=[])
        return self.distance
        
        
        
#         3
# [[0,1],[0,2]]
# [[1,0]]
        
        
# @lc code=end

File: 1129.shortest-path-with-alternating-colors.java
import java.util.ArrayList;
import java.util.Stack;
import javafx.util.Pair;  

/*
 * @lc app=leetcode id=1129 lang=java
 *
 * [1129] Shortest Path with Alternating Colors
 */

// @lc code=start
class Solution {

    class Path{
        int color;
        int directedNode;
        Path(int color, int directedNode){
            this.color = color;
            this.directedNode = directedNode;
        }
    }
    
    public ArrayList<ArrayList<Path>> makeAdjacencyList(int[][] red, int[][] blue, int num) {
        ArrayList<ArrayList<Path>> adj = new ArrayList<ArrayList<Path>>();

        for (int i = 0; i < num; i++) {
            adj.add(new ArrayList<>());
        }

        for (int[] course : red) {
            adj.get(course[0]).add(new Path(0, course[1]));
            // adj.get(course[1]).add(course[0]);

        }
        for (int[] course : blue) {
            adj.get(course[0]).add(new Path(1, course[1]));
            // adj.get(course[1]).add(course[0]);

        }

        return adj;
    }

    public int[] shortestAlternatingPaths(int n, int[][] redEdges, int[][] blueEdges) {
        ArrayList<ArrayList<Path>> graph = makeAdjacencyList(redEdges, blueEdges, n);
        int[] ans = new int[n];
        ans[0]= 0;
        Stack<Pair<Path,Integer>> bfs = new Stack<Pair<Path,Integer>>();

        bfs.push

    }
}
// @lc code=end

